### PR TITLE
Add Context.config

### PR DESCRIPTION
**Pull request by baitsguy**
Let's you set some global variables accessible in the rest of the script. The LLM and OS changes in this PR are example usages

Example code (see signature change in `write.opensearch()` and `OpenAIEntityExtractor` initialization)

```python
context = sycamore.init(
  config=Config(
      {
          Config.OPENSEARCH_CLIENT_CONFIG: os_client_args,
          Config.OPENSEARCH_INDEX_NAME: "toyindex",
          Config.OPENSEARCH_INDEX_SETTINGS: index_settings,
          Config.LLM: openai_llm,
      }
  )
)
ds = (
    context.read.binary(paths, binary_format="pdf")
    .partition(partitioner=UnstructuredPdfPartitioner())
    .extract_entity(entity_extractor=OpenAIEntityExtractor("title", prompt_template=title_context_template))
    .extract_entity(entity_extractor=OpenAIEntityExtractor("authors", prompt_template=author_context_template))
    .merge(GreedyTextElementMerger(tokenizer=tokenizer, max_tokens=300))
    .explode()
    .sketch()
    .embed(
        embedder=SentenceTransformerEmbedder(
            batch_size=100, model_name="sentence-transformers/all-MiniLM-L6-v2"
        )
    )
)
ds.write.opensearch()
```

### CODE REVIEW

File: lib/sycamore/sycamore/config.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+from typing import Dict, Optional, Any
+
+
+class Config:
+
+    OPENSEARCH_CLIENT_CONFIG = "opensearch.client_config"
+    OPENSEARCH_INDEX_NAME = "opensearch.index_name"
+    OPENSEARCH_INDEX_SETTINGS = "opensearch.index_settings"
```
Comment (HenryL27) [inline lib/sycamore/sycamore/config.py:8]: maybe these constants belong in a more opensearchy namespace?

Comment (baitsguy) [reply lib/sycamore/sycamore/config.py:8]: can reorganize as they grow, for now figured having configs in 1 place is easier to manage

File: lib/sycamore/sycamore/config.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+from typing import Optional, Any
+
+
+class Config:
```
Comment (mdwelsh) [inline lib/sycamore/sycamore/config.py:4]: Why not just use a regular `@dataclass` for this, or even a Pydantic `BaseModel` subclass (which gives you serialization and validation for free)? I don't know why one would implement this as a key-value pair mapping :-)

File: lib/sycamore/sycamore/tests/integration/connectors/opensearch/test_pdf_to_opensearch.py [hunk 3]
```diff
@@ -97,7 +98,7 @@ def test_pdf_to_opensearch_with_llm_caching():
 
         """
 
-    s3_cache_base_path = os.environ["SYCAMORE_S3_TEMP_PATH"]
+    s3_cache_base_path = os.environ.get("SYCAMORE_S3_TEMP_PATH", "s3://aryn-sycamore-integ-temp/")
```
Comment (mdwelsh) [inline lib/sycamore/sycamore/tests/integration/connectors/opensearch/test_pdf_to_opensearch.py:101]: It definitely feels like the S3 cache path should be part of the config object

File: lib/sycamore/sycamore/tests/integration/connectors/opensearch/test_pdf_to_opensearch.py [hunk 4]
```diff
@@ -133,12 +138,7 @@ def test_pdf_to_opensearch_with_llm_caching():
                 )
             )
         )
-
-        ds.write.opensearch(
-            os_client_args=os_client_args,
-            index_name="toyindex",
-            index_settings=index_settings,
-        )
+        ds.write.opensearch()
```
Comment (mdwelsh) [inline lib/sycamore/sycamore/tests/integration/connectors/opensearch/test_pdf_to_opensearch.py:141]: I tend to think that one is more likely to read from and write to different indices, so maybe the index name is still a parameter to the read.opensearch() and write.opensearch() calls, but the OS config is in the config object.

File: lib/sycamore/sycamore/writer.py [hunk 5]
```diff
@@ -97,6 +98,14 @@ def opensearch(
         from typing import Any
         import copy
 
+        if os_client_args is None:
+            os_client_args = self.context.config.get(Config.OPENSEARCH_CLIENT_CONFIG)
+        assert os_client_args is not None
+
+        if not index_name:
```
Comment (mdwelsh) [inline lib/sycamore/sycamore/writer.py:105]: See comment above, I tend to think this should be passed in on each call

File: lib/sycamore/sycamore/config.py (hunk 2 above)
Comment (baitsguy) [reply lib/sycamore/sycamore/config.py:4]: Haha fair enough, let me update

File: lib/sycamore/sycamore/tests/integration/connectors/opensearch/test_pdf_to_opensearch.py (hunk 3 above)
Comment (baitsguy) [reply lib/sycamore/sycamore/tests/integration/connectors/opensearch/test_pdf_to_opensearch.py:101]: Yeah will migrate these in follow up PRs feature wise

File: lib/sycamore/sycamore/tests/integration/connectors/opensearch/test_pdf_to_opensearch.py (hunk 4 above)
Comment (baitsguy) [reply lib/sycamore/sycamore/tests/integration/connectors/opensearch/test_pdf_to_opensearch.py:141]: They're all still parameters, they just default to reading from the config if not specified. You can still do ds.write.opensearch(index="something")

File: lib/sycamore/sycamore/writer.py (hunk 5 above)
Comment (baitsguy) [reply lib/sycamore/sycamore/writer.py:105]: replied above

File: lib/sycamore/sycamore/tests/integration/connectors/opensearch/test_pdf_to_opensearch.py (hunk 4 above)
Comment (baitsguy) [reply lib/sycamore/sycamore/tests/integration/connectors/opensearch/test_pdf_to_opensearch.py:141]: we have a couple notebooks I believe that are using the same index in a bunch of places

Comment (baitsguy) [reply lib/sycamore/sycamore/tests/integration/connectors/opensearch/test_pdf_to_opensearch.py:141]: write, then query etc

File: lib/sycamore/sycamore/config.py (hunk 2 above)
Comment (baitsguy) [reply lib/sycamore/sycamore/config.py:4]: Discussed offline, will start with a typed BaseModel and can expand it to free-formed fields if needed later

File: lib/sycamore/sycamore/context.py [hunk 6]
```diff
@@ -109,7 +117,11 @@ def deregister_rule(self, rule: Rule) -> None:
 _global_context: Optional[Context] = None
 
 
-def init(exec_mode=ExecMode.RAY, ray_args: Optional[dict[str, Any]] = None) -> Context:
+def init(exec_mode=ExecMode.RAY, ray_args: Optional[dict[str, Any]] = None, config: Optional[Config] = None) -> Context:
+    """
+    Initialized a new Context. If there is already an initialized Context, we reuse the Ray session but allow you to
+    override any Config variables.
+    """
```
Comment (HenryL27) [inline lib/sycamore/sycamore/context.py:124]: ooh actual documentation for sycamore.init()! (nit: Initialized -> Initialize; and maybe add an Args: block)

File: lib/sycamore/sycamore/context.py [hunk 7]
```diff
@@ -122,9 +134,19 @@ def init(exec_mode=ExecMode.RAY, ray_args: Optional[dict[str, Any]] = None) -> C
 
             sycamore_logger.setup_logger()
 
-            _global_context = Context(exec_mode, ray_args)
+            _global_context = Context(exec_mode, ray_args, config)
+        if config:
+            _global_context._config = config
```
Comment (HenryL27) [inline lib/sycamore/sycamore/context.py:139]: Is this gonna do funky things to my ray workers?

File: lib/sycamore/sycamore/config.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+
+@dataclass
+class Config:
+
+    opensearch_client_config: Optional[dict[str, Any]] = None
+    opensearch_index_name: Optional[str] = None
+    opensearch_index_settings: Optional[dict[str, Any]] = None
+
+    llm: Optional[LLM] = None
```
Comment (HenryL27) [inline lib/sycamore/sycamore/config.py:14]: So if I'm a user and I want to set my own config vars how do I do that? Can I just do something like:
```python
cfg = Config()
cfg.myfieldwithlikegcscredentialsorsomething = "efngsogijaeropgijw"
```
or do I have to pull subclassing shenanigans

File: lib/sycamore/sycamore/context.py (hunk 6 above)
Comment (HenryL27) [reply lib/sycamore/sycamore/context.py:124]: Maybe add this module to somewhere in the api docs toctree, idk

File: lib/sycamore/sycamore/transforms/extract_entity.py [hunk 9]
```diff
@@ -1,14 +1,14 @@
 from abc import ABC, abstractmethod
 from typing import Callable, Any, Optional, Union
 
-
+from sycamore import context
```
Comment (HenryL27) [inline lib/sycamore/sycamore/transforms/extract_entity.py:4]: this feels weird to me, not sure why. Think I'd be more comfortable with either using it as `sycamore.context.current()` or exporting current to sycamore and having `sycamore.current()`. Eh, I guess this is fine. Disregard the above unless you want to.

File: lib/sycamore/sycamore/writer.py [hunk 10]
```diff
@@ -97,6 +97,14 @@ def opensearch(
         from typing import Any
         import copy
 
+        if os_client_args is None:
+            os_client_args = self.context.config.opensearch_client_config
+        assert os_client_args is not None
```
Comment (HenryL27) [inline lib/sycamore/sycamore/writer.py:102]: Add an error message explaining what went wrong here

**Review by HenryL27 — COMMENTED** [review]
non-blocking so if you want an approve ok but idk if there are other reviewers who disagree

File: lib/sycamore/sycamore/transforms/extract_entity.py (hunk 9 above)
Comment (baitsguy) [reply lib/sycamore/sycamore/transforms/extract_entity.py:4]: I think the former is fine, little more readable too - will do

File: lib/sycamore/sycamore/writer.py (hunk 10 above)
Comment (baitsguy) [reply lib/sycamore/sycamore/writer.py:102]: Ack

**Review by HenryL27 — APPROVED** [review]

**Pull request merged**